Login: Dark mode

garrett opened the pull request:
The login page now has a dark mode which changes with your system's dark settings. Most desktops have a setting for this in "appearance" area in system settings, including GNOME, KDE, Windows, macOS, Android, and iOS/iPadOS.

Here's [Cockpit Client](https://flathub.org/apps/details/org.cockpit_project.CockpitClient), which [also recently got dark mode a few releases ago](https://cockpit-project.org/blog/cockpit-269.html), in the standard light mode:

![Screenshot from 2022-06-22 18-42-47](https://user-images.githubusercontent.com/10246/175091972-223986a2-2116-4e76-a734-5e51e36dce5d.png)

And after switching to dark settings, it looks like this:

![Screenshot from 2022-06-22 18-42-59](https://user-images.githubusercontent.com/10246/175091974-7a740111-008d-4679-bea0-7e4d1fd583c4.png)

We’re still working on adding dark mode for Cockpit once you've logged in, in the near future.

garrett commented:
TODO:

- [ ] Check in Cockpit Client
- [ ] Handle various states (hover, focus, disabled)
- [ ] Double-check color usage

garrett commented:
Screenshot in dark mode:

![Screenshot from 2022-06-01 14-22-00](https://user-images.githubusercontent.com/10246/171403335-c7f67146-eaa6-4700-9424-8524e58f0ad3.png)

garrett commented:
Here's Cockpit Client. We definitely need some more changes before merging this PR. :wink: 

![image](https://user-images.githubusercontent.com/10246/171403745-b083b7ae-e594-48eb-9734-af0f307826fe.png)

garrett commented:
There's a tiny bug with a focused link; I think I have a fix and am building a VM to test it now.

![image](https://user-images.githubusercontent.com/10246/172871926-04afb974-58d1-4e24-8494-2fd9198e7047.png)

allisonkarlitskaya commented:
Looks much better!

allisonkarlitskaya reviewed (approved):
I poked around a bit and couldn't find anything that was broken.  It looks nice!

allisonkarlitskaya reviewed (commented):
Do we also want to darken the error messages?  (the ones like in https://github.com/cockpit-project/cockpit/issues/17458).

When in dark mode, they don't change colour at all, and to my eye, seem a bit too bright, given the surrounding content.

garrett commented:
> Do we also want to darken the error messages? (the ones like in https://github.com/cockpit-project/cockpit/issues/17458).

...I did that already?

![image](https://user-images.githubusercontent.com/10246/174543635-f358d4f3-cad8-4ccc-ae5e-b3f20154d0da.png)

(This is without the overlap. I'll rebase now to get that fix.)

garrett commented:
OK, rebased; it looks like this now with the position fix:

![image](https://user-images.githubusercontent.com/10246/174544859-74d90dab-a23f-4dd0-98db-53e4dba0312d.png)

allisonkarlitskaya reviewed (approved):
Probably I was testing a previous version of the branch.  It looks good now :)

garrett commented:
Whoops. :focus and :active states on the recent hosts are problematic in dark mode:

![image](https://user-images.githubusercontent.com/10246/174546655-da2bccfb-b172-4f88-82a0-104cc07da29f.png)

I'll fix.

garrett commented:
Fixed `:focus` state

![Screenshot from 2022-06-20 11-24-17](https://user-images.githubusercontent.com/10246/174570513-c5d85b57-acf8-499c-9622-56ed72d0a3e8.png)

Additionally, I've added a focus ring to the × icon and it works in both light and dark modes. This one was tricky, as I had to move the × SVG to a `::before` as it was masking the outline. The outline has been moved to the `::after` as a result.  

![Screenshot from 2022-06-20 11-24-26](https://user-images.githubusercontent.com/10246/174570519-a90d4891-15bb-4405-a7d2-f3bcf52dbb55.png)

![Screenshot from 2022-06-20 11-24-34](https://user-images.githubusercontent.com/10246/174570521-543d796c-5dd0-4a85-9fc0-f47cb8ffe887.png)

garrett commented:
I realized I didn't need both ::before and ::after. I could use the ::before for the × and the main element for the focus ring. Adjusted CSS appropriately @ https://github.com/cockpit-project/cockpit/compare/9660b3bda7101469aeb110e95e80f6f1c2d4560a..8f72d70d47500544d563e216d36cba00de244a90.

jelly commented on pkg/static/login.css:
```diff
@@ -922,6 +929,112 @@ input[type=password] + .login-password-toggle .password-hide {
+  a:active,
+  a:focus,
+  a:hover {
+    color:#8ac0f6;
```
Should be a space after `:`, minor nitpick.

garrett replied:
Yeah, not sure how this happened. I guess I copy/pasted and it dropped the space or something. Whoops.

jelly commented on pkg/static/login.css:
```diff
@@ -922,6 +929,112 @@ input[type=password] + .login-password-toggle .password-hide {
+  .pf-m-tertiary {
+    background-color: transparent;
+    border-color:#8ac0f6;
+    color:#8ac0f6;
```
space after `:`, minor nitpick.

jelly commented on pkg/static/login.css:
```diff
@@ -922,6 +929,112 @@ input[type=password] + .login-password-toggle .password-hide {
+
+  .pf-c-alert.pf-m-danger > svg,
+  .pf-c-alert.pf-m-danger .pf-c-alert__title {
+    color: #f64747;
```
It seems that patternfly uses a slightly different color for this one:

```
[jelle@t14s][~/projects/patternfly]%git grep pf-color-red-9999;
src/patternfly/base/themes/dark/_variables.scss:  --pf-global--palette--red-9999: #{$pf-color-red-9999};
src/patternfly/sass-utilities/themes/dark/colors.scss:$pf-color-red-9999: #fe5142;
src/patternfly/sass-utilities/themes/dark/scss-variables.scss:$pf-global--danger-color--100: $pf-color-red-9999;
```

It is more orange'ish somehow :)

garrett replied:
I pulled the colors directly from their mockups. They might've changed it?

jelly commented:
Looks good on an unsupported browser with dark theme 

![image](https://user-images.githubusercontent.com/67428/174626453-7bfec2d0-02d6-48d8-9f55-32d2916eadf2.png)

jelly commented:
Really small nitpick, does the spinner look how it's supposed too?

![image](https://user-images.githubusercontent.com/67428/174630044-b796db63-d275-4802-a9ca-e185a2a7af04.png)

garrett commented:
@jelly: I've resolved your issues and found a few more things to improve while I was there.

Changes since your last look:

https://github.com/cockpit-project/cockpit/compare/8f72d70d47500544d563e216d36cba00de244a90..fc72a0ec88f88fd666270e131e4fabbce1120d8b

marusak commented:
```
ERROR in [stylelint] 
pkg/static/login.css
 1024:5  ✖  Unexpected unknown property "text-color"  property-no-unknown
```

garrett commented on pkg/static/login.css:
```diff
@@ -922,6 +929,118 @@ input[type=password] + .login-password-toggle .password-hide {
+  .pf-c-button[disabled] {
+    background-color: #444548;
+    border-color: #444548;
+    text-color: #c6c7c8
```
```suggestion
    color: #c6c7c8
```

allisonkarlitskaya reviewed (approved):
It is time.
